# Predicting Student Health Risk

## Data Understanding

This notebook is used to load and inspect the official Kaggle competition datasets before data cleaning, preprocessing, and model development.

In [28]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pathlib import Path

print("Required libraries imported successfully")

Required libraries imported successfully


In [29]:
current_folder = Path.cwd()

print("Current working folder")
print(current_folder)

Current working folder
c:\Users\Admin\Documents\Rakindu's Files\ICBT\Computational Intelligence\Assignment\Student health risk project\notebooks


In [30]:
if (current_folder / "data").exists():
    project_folder = current_folder
else:
    project_folder = current_folder.parent

print("Project folder")
print(project_folder)

Project folder
c:\Users\Admin\Documents\Rakindu's Files\ICBT\Computational Intelligence\Assignment\Student health risk project


In [31]:
data_folder = project_folder / "data" / "kaggle" / "raw"

print("Dataset folder")
print(data_folder)

print("Folder exists")
print(data_folder.exists())

Dataset folder
c:\Users\Admin\Documents\Rakindu's Files\ICBT\Computational Intelligence\Assignment\Student health risk project\data\kaggle\raw
Folder exists
True


In [32]:
dataset_files = list(data_folder.iterdir())

print("Files inside the dataset folder")

for file in dataset_files:
    print(file.name)

Files inside the dataset folder
sample_submission.csv
test.csv
train.csv


In [33]:
train_file = data_folder / "train.csv"
test_file = data_folder / "test.csv"
submission_file = data_folder / "sample_submission.csv"

print("Training file exists", train_file.exists())
print("Testing file exists", test_file.exists())
print("Submission file exists", submission_file.exists())

Training file exists True
Testing file exists True
Submission file exists True


In [34]:
train_data = pd.read_csv(train_file)
test_data = pd.read_csv(test_file)
sample_submission = pd.read_csv(submission_file)

print("All datasets loaded successfully")

All datasets loaded successfully


In [35]:
print("Training dataset shape", train_data.shape)
print("Testing dataset shape", test_data.shape)
print("Sample submission shape", sample_submission.shape)

Training dataset shape (690088, 15)
Testing dataset shape (295753, 14)
Sample submission shape (295753, 2)


In [36]:
train_data.head()

,id,health_condition,sleep_duration,heart_rate,bmi,calorie_expenditure,step_count,exercise_duration,water_intake,diet_type,stress_level,sleep_quality,physical_activity_level,smoking_alcohol,gender
0,0,unhealthy,5.22,70.6,25.66,2174.0,1326.0,19.8,1.86,veg,high,average,sedentary,yes,female
1,1,at-risk,5.53,71.3,25.84,1966.0,9891.0,49.9,1.26,non-veg,low,average,moderate,yes,other
2,2,unhealthy,5.29,75.4,24.54,2688.0,14216.0,38.1,1.60,veg,high,poor,active,yes,male
3,3,unhealthy,4.70,77.2,23.13,2630.0,7174.0,59.9,2.02,veg,high,average,active,occasional,female
4,4,at-risk,7.23,73.4,28.44,2560.0,6584.0,46.0,2.25,veg,NaN,average,sedentary,NaN,male


In [37]:
test_data.head()

,id,sleep_duration,heart_rate,bmi,calorie_expenditure,step_count,exercise_duration,water_intake,diet_type,stress_level,sleep_quality,physical_activity_level,smoking_alcohol,gender
0,690088,5.35,64.9,23.48,2745.0,14167.0,59.5,1.86,veg,high,poor,active,occasional,male
1,690089,NaN,83.1,22.42,1773.0,6801.0,24.5,2.40,balanced,high,poor,sedentary,yes,other
2,690090,6.68,59.7,24.14,3040.0,13250.0,48.5,2.76,balanced,medium,poor,active,no,NaN
3,690091,7.13,78.5,26.26,2494.0,6331.0,56.9,2.34,veg,low,good,moderate,yes,other
4,690092,5.49,77.7,23.29,1828.0,13894.0,39.4,2.45,veg,high,average,active,occasional,other


In [38]:
sample_submission.head()

,id,health_condition
0,690088,at-risk
1,690089,at-risk
2,690090,at-risk
3,690091,at-risk
4,690092,at-risk


In [39]:
print("Training columns")

for column in train_data.columns:
    print(column)

print()

print("Testing columns")

for column in test_data.columns:
    print(column)

Training columns
id
health_condition
sleep_duration
heart_rate
bmi
calorie_expenditure
step_count
exercise_duration
water_intake
diet_type
stress_level
sleep_quality
physical_activity_level
smoking_alcohol
gender

Testing columns
id
sleep_duration
heart_rate
bmi
calorie_expenditure
step_count
exercise_duration
water_intake
diet_type
stress_level
sleep_quality
physical_activity_level
smoking_alcohol
gender


In [40]:
target_column = "health_condition"

print("Target column", target_column)
print("Target exists in training data", target_column in train_data.columns)
print("Target exists in testing data", target_column in test_data.columns)

Target column health_condition
Target exists in training data True
Target exists in testing data False


In [41]:
train_data.dtypes

id                           int64
health_condition               str
sleep_duration             float64
heart_rate                 float64
bmi                        float64
calorie_expenditure        float64
step_count                 float64
exercise_duration          float64
water_intake               float64
diet_type                      str
stress_level                   str
sleep_quality                  str
physical_activity_level        str
smoking_alcohol                str
gender                         str
dtype: object

In [42]:
train_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 690088 entries, 0 to 690087
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   id                       690088 non-null  int64  
 1   health_condition         690088 non-null  str    
 2   sleep_duration           614089 non-null  float64
 3   heart_rate               682255 non-null  float64
 4   bmi                      676190 non-null  float64
 5   calorie_expenditure      637235 non-null  float64
 6   step_count               676172 non-null  float64
 7   exercise_duration        683187 non-null  float64
 8   water_intake             646611 non-null  float64
 9   diet_type                683187 non-null  str    
 10  stress_level             607277 non-null  str    
 11  sleep_quality            631757 non-null  str    
 12  physical_activity_level  653467 non-null  str    
 13  smoking_alcohol          661506 non-null  str    
 14  gender         

In [43]:
test_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 295753 entries, 0 to 295752
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   id                       295753 non-null  int64  
 1   sleep_duration           263182 non-null  float64
 2   heart_rate               292396 non-null  float64
 3   bmi                      289797 non-null  float64
 4   calorie_expenditure      273101 non-null  float64
 5   step_count               289789 non-null  float64
 6   exercise_duration        292795 non-null  float64
 7   water_intake             277120 non-null  float64
 8   diet_type                292795 non-null  str    
 9   stress_level             260263 non-null  str    
 10  sleep_quality            270754 non-null  str    
 11  physical_activity_level  280058 non-null  str    
 12  smoking_alcohol          283504 non-null  str    
 13  gender                   286593 non-null  str    
dtypes: float64(7), 

In [44]:
train_missing_values = train_data.isnull().sum()

print("Missing values in training data")
print(train_missing_values)

Missing values in training data
id                             0
health_condition               0
sleep_duration             75999
heart_rate                  7833
bmi                        13898
calorie_expenditure        52853
step_count                 13916
exercise_duration           6901
water_intake               43477
diet_type                   6901
stress_level               82811
sleep_quality              58331
physical_activity_level    36621
smoking_alcohol            28582
gender                     21373
dtype: int64


In [45]:
test_missing_values = test_data.isnull().sum()

print("Missing values in testing data")
print(test_missing_values)

Missing values in testing data
id                             0
sleep_duration             32571
heart_rate                  3357
bmi                         5956
calorie_expenditure        22652
step_count                  5964
exercise_duration           2958
water_intake               18633
diet_type                   2958
stress_level               35490
sleep_quality              24999
physical_activity_level    15695
smoking_alcohol            12249
gender                      9160
dtype: int64


In [46]:
train_missing_percentage = train_data.isnull().mean() * 100
test_missing_percentage = test_data.isnull().mean() * 100

print("Training missing-value percentages")
print(train_missing_percentage.round(2))

print()

print("Testing missing-value percentages")
print(test_missing_percentage.round(2))

Training missing-value percentages
id                          0.00
health_condition            0.00
sleep_duration             11.01
heart_rate                  1.14
bmi                         2.01
calorie_expenditure         7.66
step_count                  2.02
exercise_duration           1.00
water_intake                6.30
diet_type                   1.00
stress_level               12.00
sleep_quality               8.45
physical_activity_level     5.31
smoking_alcohol             4.14
gender                      3.10
dtype: float64

Testing missing-value percentages
id                          0.00
sleep_duration             11.01
heart_rate                  1.14
bmi                         2.01
calorie_expenditure         7.66
step_count                  2.02
exercise_duration           1.00
water_intake                6.30
diet_type                   1.00
stress_level               12.00
sleep_quality               8.45
physical_activity_level     5.31
smoking_alcohol         

In [47]:
training_duplicates = train_data.duplicated().sum()
testing_duplicates = test_data.duplicated().sum()

print("Duplicated training rows", training_duplicates)
print("Duplicated testing rows", testing_duplicates)

Duplicated training rows 0
Duplicated testing rows 0


In [48]:
target_counts = train_data[target_column].value_counts()

print("Target class counts")
print(target_counts)

Target class counts
health_condition
at-risk      592561
unhealthy     57724
fit           39803
Name: count, dtype: int64


In [49]:
target_percentages = train_data[target_column].value_counts(normalize=True) * 100

print("Target class percentages")
print(target_percentages.round(2))

Target class percentages
health_condition
at-risk      85.87
unhealthy     8.36
fit           5.77
Name: proportion, dtype: float64


In [50]:
numerical_columns = train_data.select_dtypes(
    include=["int64", "float64"]
).columns.tolist()

categorical_columns = train_data.select_dtypes(
    include=["object"]
).columns.tolist()

print("Numerical columns")
print(numerical_columns)

print()

print("Categorical columns")
print(categorical_columns)

Numerical columns
['id', 'sleep_duration', 'heart_rate', 'bmi', 'calorie_expenditure', 'step_count', 'exercise_duration', 'water_intake']

Categorical columns
['health_condition', 'diet_type', 'stress_level', 'sleep_quality', 'physical_activity_level', 'smoking_alcohol', 'gender']


C:\Users\Admin\AppData\Local\Temp\ipykernel_18976\3952475703.py:5: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_columns = train_data.select_dtypes(


In [51]:
categorical_feature_columns = categorical_columns.copy()

if target_column in categorical_feature_columns:
    categorical_feature_columns.remove(target_column)

print("Categorical feature columns")
print(categorical_feature_columns)

Categorical feature columns
['diet_type', 'stress_level', 'sleep_quality', 'physical_activity_level', 'smoking_alcohol', 'gender']


In [52]:
train_data.describe().T

,count,mean,std,min,25%,50%,75%,max
id,690088.0,345043.500000,199211.390620,0.0,172521.75,345043.50,517565.25,690087.00
sleep_duration,614089.0,6.992597,1.215407,3.0,6.16,6.99,7.81,10.00
heart_rate,682255.0,75.096504,8.175106,50.0,69.40,75.10,80.70,107.70
bmi,676190.0,22.984925,2.481787,16.0,21.32,22.99,24.66,34.82
calorie_expenditure,637235.0,2226.084931,347.532098,1200.0,2053.00,2241.00,2456.00,3580.00
step_count,676172.0,8615.953050,3929.399831,1002.0,5389.00,8856.00,12114.00,14999.00
exercise_duration,683187.0,38.751456,14.742189,0.0,29.20,39.40,49.40,99.80
water_intake,646611.0,2.188542,0.518489,0.5,1.84,2.17,2.50,4.72


In [53]:
train_data.describe(include="object").T

C:\Users\Admin\AppData\Local\Temp\ipykernel_18976\1655501410.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  train_data.describe(include="object").T


,count,unique,top,freq
health_condition,690088,3,at-risk,592561
diet_type,683187,3,veg,231432
stress_level,607277,3,medium,261819
sleep_quality,631757,3,average,213948
physical_activity_level,653467,3,moderate,221041
smoking_alcohol,661506,3,yes,223730
gender,668715,3,male,237756


In [54]:
training_features = set(train_data.columns) - {target_column}
testing_features = set(test_data.columns)

missing_from_test = training_features - testing_features
extra_in_test = testing_features - training_features

print("Training features missing from testing data")
print(missing_from_test)

print()

print("Extra features in testing data")
print(extra_in_test)

Training features missing from testing data
set()

Extra features in testing data
set()


## Initial observations

The training dataset contains 690,088 rows and 15 columns.

The testing dataset contains 295,753 rows and 14 columns.

The target variable is `health_condition`.

This is a supervised multiclass classification problem with the classes `at-risk`, `unhealthy`, and `fit`.

The raw training dataset contains 8 numerical columns including `id` and 6 categorical input features.

The training dataset contains 449,496 missing values and the testing dataset contains 192,642 missing values.

No duplicated rows were found in either dataset.

The target is highly imbalanced because `at-risk` represents 85.87 percent, `unhealthy` represents 8.36 percent, and `fit` represents 5.77 percent.

The next stage is Exploratory Data Analysis.